File: js/GraphAdjMatrix.js
export class Graph {

	constructor(directed = true) {
		this.directed = directed
		this.matrix = []
		this.nodes = []
	}
	addNode(id) {
		this.nodes.push(id)
		for(let line of this.matrix) {
			line.push(null)
		}
		this.matrix.push(Array(this.nodes.length).fill(null))
	}
	addEdge(source, target, weight=1) {
		const sourceIndex = this.nodes.indexOf(source)
		const targetIndex = this.nodes.indexOf(target)
		this.matrix[sourceIndex][targetIndex] = weight
		if(!this.directed) {
			this.matrix[targetIndex][sourceIndex] = weight
		}
	}
	totalNodes() {return this.nodes.length}
	getNeighbours(id) {
		const neighbours = []
		const index = this.nodes.indexOf(id)
		for(let i = 0; i < this.matrix[index].length; i++) {
			const weight = this.matrix[index][i]
			if(weight !== null) {
				neighbours.push({id: this.nodes[i], weight})
			}
		}
		return neighbours
	}
	totalNodes() {return this.nodes.length}
	getEdges() {
		const edges = []
		for(let i = 0; i < this.matrix.length; i++) {
			for(let j = 0; j < this.matrix[i].length; j++) {
				const weight = this.matrix[i][j]
				if(weight !== null) {
					edges.push({source: this.nodes[i], target: this.nodes[j], weight})
				}
			}
		}
		return this.edges
	}
	getNodes() {return this.nodes}
}

File: python/GraphAdjMatrix.py
class Graph:

	def __init__(self, directed = True):
		self.directed = directed
		self.matrix = []
		self.nodes = []

	def addNode(self, id):
		self.nodes.append(id)
		for line in self.matrix:
			line.append(None)

		self.matrix.append([None] * len(self.nodes))

	def addEdge(self, source, target, weight=1):
		sourceIndex = self.nodes.index(source)
		targetIndex = self.nodes.index(target)
		self.matrix[sourceIndex][targetIndex] = weight
		if(not self.directed):
			self.matrix[targetIndex][sourceIndex] = weight

	def totalNodes(self):
		return len(self.nodes)


	def getNeighbours(self, id):
		neighbours = []
		index = self.nodes.index(id)
		for i in range(len(self.matrix[index])): 
			weight = self.matrix[index][i]
			if(weight != None):
				neighbours.append({"id": self.nodes[i], "weight": weight})

		return neighbours
